### PR TITLE
Grant permissions to kiali operator SA to read Secrets

### PR DESCRIPTION
**Pull request by leandroberetta**
Needed by [#4227](https://github.com/kiali/kiali/pull/4227)

### CODE REVIEW

**Review by jmazzitelli — CHANGES_REQUESTED** [review]
This is a problem. We are going to need a project-wide discussion before we merge. Do not merge until this is vetted by the team and community.

We specifically have been avoiding adding this "read any secret" permission. People do not like it (we recently had a community member complain about our current ability to read secrets in the control plane namepace). This is a very, very bad (potential) security hold that kiali would be opening up.

We need to avoid this permission. If you need to read a specific secret, we can do so by name as we do here: https://github.com/kiali/helm-charts/blob/master/kiali-operator/templates/clusterrole.yaml#L41-L52

File: kiali-operator/templates/clusterrole.yaml
Comment (jmazzitelli) [inline kiali-operator/templates/clusterrole.yaml:188]: this is a no-op change (spaces at EOL?) - revert
```suggestion
  resources:
```

File: kiali-server/templates/role-controlplane.yaml
Comment (jmazzitelli) [inline kiali-server/templates/role-controlplane.yaml:26]: This section isn't using the new config setting. This instead should be:
```suggestion
{{- if .Values.kiali_feature_flags.certificates_information_indicators.enabled }}
- apiGroups: [""]
  resourceNames: {{ .Values.kiali_feature_flags.certificates_information_indicators.secrets }}
  resources:
  - secrets
  verbs:
  - get
  - list
  - watch
{{- end }}
```
I'm not exactly sure of the syntax to get that list value set in `resourceNames` but you get the idea.

**Review by jmazzitelli — CHANGES_REQUESTED** [review]

File: kiali-operator/templates/clusterrole.yaml
Comment (jmazzitelli) [inline kiali-operator/templates/clusterrole.yaml:38]: see my comment coming about `secretReader: true` setting

**Comment by jmazzitelli** [comment]
I think we want to add another flag to let people turn off the global "get" permission if they don't want to even grant the operator that permission.

So we want an analogous setting as this: https://github.com/kiali/helm-charts/blob/v1.38.0/kiali-operator/values.yaml#L35-L38
which is used here: https://github.com/kiali/helm-charts/blob/v1.38.0/kiali-operator/templates/clusterrole.yaml#L102-L105

For example:

values.yaml:
```
# Set to true if you want the operator to be able to read secrets across the cluster. This is necessary
# if you want to support Kiali CRs with spec.kiali_feature_flags.certificates_information_indicators.enabled = true.
# If you set this to false, the operator will not be given permission to read secrets outside of the control plane namespace.
secretReader: true
```

templates/clusterrole.yaml:
```
  {{- if .Values.secretReader }}
  - get
  {{- end }}
```

This should help users that are paranoid and don't want to even the operator to be granted permissions to read all secrets. I don't think the operator has global/cluster-wide ability to read secrets today - this keeps it that way if the user wants.

**Comment by jmazzitelli** [comment]
I created PR https://github.com/leandroberetta/helm-charts/pull/1 that adds the stuff I talk about above. This makes me feel alittle better :)

File: kiali-server/values.yaml
Comment (jmazzitelli) [inline kiali-server/values.yaml:80]: can you swap this to sort them alphabetically?
It's stupid, I know - but when we do diff's on these lists, its nice when things are sorted because the diffs won't just show different orders of things.
so this is trivial:
```suggestion
   - cacerts
   - istio-ca-secret
```

**Review by jmazzitelli — CHANGES_REQUESTED** [review]
trivial change requested

File: kiali-operator/values.yaml
Comment (jmazzitelli) [inline kiali-operator/values.yaml:42]: I would like to reword this now that I'm looking at it and reading what I wrote. That last sentence isn't accurate because the operator will still have permission to read a secret named "kiali-signing-key" found anywhere. And it isn't true that the operator can't read secrets "outside of the control plane namespace' because the operator doesn't know ahead of time what control  plane namespaces there are or will be. So let's reword this to make it more clear and specific as to what this setting does:

```suggestion
# Set to a list of secrets in the cluster that the operator will be allowed to read. This is necessary if you want to
# support Kiali CRs with spec.kiali_feature_flags.certificates_information_indicators.enabled=true.
# The secrets in this list will be the only ones allowed to be specified in any Kiali CR (in the setting
# spec.kiali_feature_flags.certificates_information_indicators.secrets).
# If you set this to an empty list, the operator will not be given permission to read any additional secrets
# found in the cluster, and thus will only support a value of "false" in the Kiali CR setting
# spec.kiali_feature_flags.certificates_information_indicators.enabled.
```

**Review by jmazzitelli — CHANGES_REQUESTED** [review]
We need to update the comment for secretReader to make it more clear what that is.

File: kiali-server/templates/role-controlplane.yaml
Comment (jmazzitelli) [inline kiali-server/templates/role-controlplane.yaml:30]: This is in the wrong spot. If the feature flags.enabled is false, then the end of file "..." marker isn't included. We want that included all the time, whether the enabled flag is true or not.

So make the final two lines of this file:

```
{{- end }}
...
```

**Review by jmazzitelli — DISMISSED** [review]
sorry - just noticed one more thing wrong.

File: kiali-server/templates/role-controlplane.yaml
Comment (leandroberetta) [reply kiali-server/templates/role-controlplane.yaml:29]: Right, thanks, fixed.

**Comment by jmazzitelli** [comment]
I just tested by setting `--set "secretReader={cacerts}"` - and things worked as expected. The default Kiali CR settings cause a failure because kiali can't be granted access to read istio-ca-secret. However, when I then decided to not give the operator any get-secret perms, helm spit out an error and the operator did not get installed:

```
$ helm install --set "secretReader={}" --set "image.repo=localhost:5000/kiali/kiali-operator" --set image.tag=dev --set allowAdHocKialiImage=true -n kiali-operator kiali-operator _output/charts/kiali-operator-1.39.0-SNAPSHOT.tgz 

Error: unable to build kubernetes objects from release manifest: error validating "": error validating data: ValidationError(ClusterRole.rules[3].resourceNames): unknown object type "nil" in ClusterRole.rules[3].resourceNames[0]
```

I am investigating why that happened.

**Review by jmazzitelli — DISMISSED** [review]
after a bunch of testing, this all looks good.

**Review by jmazzitelli — APPROVED** [review]
re-approve

**Pull request merged**
